cpp/src/quantiles/groupby.cu

```diff
@@ -0,0 +1,269 @@
+  auto sorted_order = static_cast<gdf_size_type*>(key_sort_order().data);
+  decltype(_group_offsets->begin()) result_end;
+  auto exec = rmm::exec_policy(_stream)->on(_stream);
+
+  if (has_nulls(_keys)) {
+    auto comp = row_equality_comparator<true>(*device_input_table, true);
+    result_end = thrust::unique_copy(exec,
+      thrust::make_counting_iterator<gdf_size_type>(0),
+      thrust::make_counting_iterator<gdf_size_type>(num_keys()),
+      _group_offsets->begin(), transform_row_eq_comparator<true>{comp, sorted_order});
+  } else {
+    auto comp = row_equality_comparator<false>(*device_input_table, true);
+    result_end = thrust::unique_copy(exec, 
+      thrust::make_counting_iterator<gdf_size_type>(0),
+      thrust::make_counting_iterator<gdf_size_type>(num_keys()),
+      _group_offsets->begin(), transform_row_eq_comparator<false>{comp, sorted_order});
+  }
+
+  gdf_size_type num_groups = thrust::distance(_group_offsets->begin(), result_end);
+  _group_offsets->resize(num_groups);
+
+  return *_group_offsets;
+}
+
+rmm::device_vector<gdf_size_type> const& groupby::group_labels() {
+  if (_group_labels)
+    return *_group_labels;
+
+  // Get group labels for future use in segmented sorting
+  _group_labels = std::make_unique<index_vector>(num_keys());
+
+  auto& group_labels = *_group_labels;
+  auto exec = rmm::exec_policy(_stream)->on(_stream);
+  auto group_labels_ptr = group_labels.data().get();
+  auto group_offsets_ptr = group_offsets().data().get();
+  thrust::for_each_n(exec,
+    thrust::make_counting_iterator(1),
+    num_groups() - 1,
+    [group_labels_ptr, group_offsets_ptr] __device__ (gdf_size_type i) {
+      group_labels_ptr[group_offsets_ptr[i]] = 1;
+    });
+  thrust::inclusive_scan(exec,
+                        group_labels.begin(),
+                        group_labels.end(),
+                        group_labels.begin());
+
+  return group_labels;
+}
+
+gdf_column const& groupby::unsorted_labels() {
+  if (_unsorted_labels)
+    return *_unsorted_labels;
+
+  _unsorted_labels = gdf_col_pointer(
+    new gdf_column(
+      allocate_column(gdf_dtype_of<gdf_size_type>(),
+                      key_sort_order().size,
+                      true,
+                      gdf_dtype_extra_info{},
+                      _stream)),
+    [](gdf_column* col) { gdf_column_free(col); });
+
+  CUDA_TRY(cudaMemsetAsync(_unsorted_labels->valid, 0,
+                           gdf_num_bitmask_elements(_unsorted_labels->size), 
+                           _stream));
+  
+  gdf_column group_labels_col{};
+  gdf_column_view(&group_labels_col, 
+                  const_cast<gdf_size_type*>(group_labels().data().get()), 
+                  nullptr,
+                  group_labels().size(), 
+                  gdf_dtype_of<gdf_size_type>());
+  cudf::table t_sorted_labels{&group_labels_col};
+  cudf::table t_unsorted_labels{_unsorted_labels.get()};
+  cudf::detail::scatter(&t_sorted_labels,
```

Comment:
It seems as though using `libcudf`'s scatter here is unnecessary and it's possible to just use `thrust::scatter` directly. 

The only reason to use `cudf::detail::scatter` is if data you're scattering can contain nulls, which in this case `group_labels` is non-nullable. 

So this whole function could be simplified to just:
```
index_vector unsorted_labels(num_rows());
thrust::scatter(exec, group_labels().begin(), group_labels().end(), static_cast<gdf_size_type*>(key_sort_order().data), unsorted_labels.begin());
return unsorted_labels;
```

Reply:
`unsorted_labels` is nullable. I use `cudf::scatter` to set its nulls to be same as nulls in keys. I didn't store and apply `row_bitmask` to it is because that'd be ugly and not how abstractions should work.

Reply:
If `unsorted_labels` can be nullable, then you've absolutely taken the correct approach. 

However, I'm a bit confused about how `unsorted_lables` can contain null values, especially when `group_labels` _cannot_ contain nulls. My understanding (and based on the naming/documentation) is that the only difference between the two is the ordering.

Reply:
`unsorted_labels` is a stand-in for the key table. Its size is the same as the number of input keys. `group_labels` is only of the size of `num_keys()`, which is the effective number of keys (nulls removed). `group_labels` are of the same size as a sorted value column, which also has the values removed from it where the keys are null.

Reply:
I though of `key_labels`, `key_group_labels`. But they don't include the fact that we mean the labels of unsorted keys. How about `unsorted_keys_labels`?